modules/personalize_taxonomy_context/plugins/visitor_context/TaxonomyContext.inc

```diff
@@ -0,0 +1,93 @@
+<?php
+
+/**
+ * @file
+ * Provides a visitor context plugin for taxonomy.
+
+ */
+
+class TaxonomyContext extends PersonalizeContextBase {
+  /**
+   * Implements PersonalizeContextInterface::create().
+   */
+  public static function create(PersonalizeAgentInterface $agent = NULL, $selected_context = array()) {
+    return new self($agent, $selected_context);
+  }
+
+  /**
+   * Implements PersonalizeContextInterface::getOptions().
+   */
+  public static function getOptions() {
+    $vocabularies = taxonomy_get_vocabularies();
+    $options = array();
+    foreach ($vocabularies as $vocabulary) {
+      $machine_name = $vocabulary->machine_name;
+      $options[$machine_name] = array(
+        'name' => $vocabulary->name,
+        'group' => 'Taxonomy',
+        'id' => $vocabulary->vid,
+      );
+    }
+    return $options;
+  }
+
+  /**
+   * Implements PersonalizeContextInterface::getPossibleValues().
+   */
+  public function getPossibleValues($limit = FALSE) {
+    $possible_values = array();
+    $options = $this->getOptions();
+    $free_tagging_vocabulary_machine_names = SELF::getFreeTaggingVocabulariesMachineNames();
+    foreach ($options as $machine_name => $info) {
+      // If the vocabulary is free-tagging, then we don't use it for lift targeting.
+      if (in_array($machine_name, $free_tagging_vocabulary_machine_names)) {
+        continue;
+      }
+      $values = array();
+      $terms = taxonomy_get_tree($info['id']);
+      foreach($terms as $term) {
+        $values[$term->tid] = $term->name;
+      }
+
+      $name = $info['name'];
+      $possible_values[$name] = array(
+        'value type' => 'predefined',
+        'values' => $values,
+      );
+    }
+    return $possible_values;
+  }
+
+  /**
+   * Implements PersonalizeAgentInterface::getAssets().
+   */
+  public function getAssets() {
+    return array(
+      'js' => array(
+        drupal_get_path('module', 'personalize_taxonomy_context') . '/js/personalize_taxonomy_context.js' => array(),
+        array(
+          'data' => array('personalize_taxonomy_context' => personalize_taxonomy_context_node_taxonomy()),
```

Comment:
In fact, it was coded this way because I thought we *can* rely on the fact that the static variable is already populated. This is because getAssets() is only ever called by a function ```implement hook_page_build```, and ```hook_page_build``` happens after ```hook_node_view```. _(citation needed)_